flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/QueryOperation.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.operations;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.util.StringUtils;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Base class for representing an operation structure behind a user-facing {@link Table} API.
+ *
+ * <p>It represents an operation that can be a node of a relational query. It has a schema, that
+ * can be used to validate a {@link QueryOperation} applied on top of this one.
+ */
+@Internal
+public abstract class QueryOperation implements TableOperation {
```

Comment:
How about we further cataloguing all the query-lated sub classes into like `LeafQueryOperation`, `UnaryQueryOperation` and `BinaryQueryOperation`, just like the original `LogicalNode` did. I think this will make user easier to understand this class family.

Reply:
To be honest, I think introducing the `Unary` leaf etc. makes the hierarchy unnecessary more complicated. The only benefit we get are default implementations of `getChildren` method.